Backward simplication

Rocky640 opened the pull request:
Rewrite of Stefan Geschwentner idea for a more natural backward definition which led to a small simplification to the master code.

On top of the usual conditions 
a) some opponent in front (but no lever)
b) some neighbours (in front) (but no neighbour behind or same rank)
c) < rank_5

to find out if a pawn is backward we look at the squares in front of this pawn to reach the same rank as the next neighbour.

In current master, a pawn is backward if any of those squares is controlled by an enemy pawn on an adjacent file

In this version, a pawn is ALSO backward if any of those squares is occupied by an enemy pawn.

STC: 
http://tests.stockfishchess.org/tests/view/56fe7efd0ebc59301a3541f1
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 19051 W: 3557 L: 3433 D: 12061

LTC:
http://tests.stockfishchess.org/tests/view/56febc2d0ebc59301a354209 
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 40810 W: 5619 L: 5526 D: 29665

mcostalba commented on src/pawns.cpp:
```diff
@@ -151,21 +151,21 @@ namespace {
-            backward = (b | shift_bb<Up>(b)) & theirPawns;
+            // backward because it is opposed or it can be captured before 
+            // making a phalanx with closest neighbour.
+            backward = (b | shift_bb<Up>(b & adjacent_files_bb(f))) & stoppers;
```
Probably this 'b & adjacent_files_bb' needs some comment...

Rocky640 commented:
OK, I simplified even further, (no functional change) and updated the comments
https://github.com/Rocky640/Stockfish/commit/d02c23a430b1e1b541066b4965a408e760c1ef6b

lucasart commented:
Nice.

lucasart commented on src/pawns.cpp:
```diff
@@ -139,33 +139,25 @@ namespace {
-        if (   (passed | isolated | lever | connected)
-            || (ourPawns & pawn_attack_span(Them, s))
-            || (relative_rank(Us, s) >= RANK_5))
+        if ( (isolated | lever | connected) || (relative_rank(Us, s) >= RANK_5))
```
Can you use the same indentation as in the original code ?

mcostalba commented on src/pawns.cpp:
```diff
@@ -131,7 +131,7 @@ namespace {
         doubled     =   ourPawns   & forward_bb(Us, s);
         opposed     =   theirPawns & forward_bb(Us, s);
-        passed      = !(theirPawns & passed_pawn_mask(Us, s));
+        stoppers    =   theirPawns & passed_pawn_mask(Us, s);
```
Sorry to be nitpicking, but while there could you pelase reindent all?

```
        // Flag the pawn
        neighbours = ourPawns   & adjacent_files_bb(f);
        doubled    = ourPawns   & forward_bb(Us, s);
        opposed    = heirPawns  & forward_bb(Us, s);
        stoppers   = theirPawns & passed_pawn_mask(Us, s);
        lever      = theirPawns & pawnAttacksBB[s];
        phalanx    = neighbours & rank_bb(s);
        supported  = neighbours & rank_bb(s - Up);
        connected  = supported | phalanx;
        isolated   = !neighbours;
```

mcostalba commented on src/pawns.cpp:
```diff
@@ -139,33 +139,25 @@ namespace {
         }
 
-        assert(opposed | passed | (pawn_attack_span(Us, s) & theirPawns));
+        assert(opposed | !stoppers | (pawn_attack_span(Us, s) & theirPawns));
```
This assert is wrong because !stoppers is a bool and is bitwised 'or' with bitboards.

mcostalba commented:
I have added a further tidy-up patch above your work: https://github.com/mcostalba/Stockfish/commit/backward2

joergoster commented:
Well done!

Rocky640 commented:
OK, I copy/pasted Marco's pawn.cpp.

I just wonder. Using || instead of | will do more branching isn't it ?
if (!neighbours || lever || connected || relative_rank(Us, s) >= RANK_5) 

If we keep the ||, and if order is important  my educated guess would be 
if (connected || lever || !neighbours || relative_rank(Us, s) >= RANK_5)
but let me run some debug_hit later today.

mcostalba commented:
Yes, debug_hit() is the way to go, sorry for not having run it myself.

P.S: I moved from | to || because we are mixing booleans and bitboards,
although technically it works, it is anyhow a bit too tricky from my taste.




On Mon, Apr 4, 2016 at 2:02 PM, Rocky640 <notifications@github.com> wrote:

> OK, I copy/pasted Marco's pawn.cpp.
>
> I just wonder. Using || instead of | will do more branching isn't it ?
> if (!neighbours || lever || connected || relative_rank(Us, s) >= RANK_5)
>
> If we keep the ||, and if order is important my educated guess would be
> if (connected || lever || !neighbours || relative_rank(Us, s) >= RANK_5)
> but let me run some debug_hit later today.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/614#issuecomment-205264597>
>

Rocky640 commented:
OK just push another version.

Ah ah, too bad for the educated guess,
The optimal order, based only on a "bench" run is
if (connected ||  !neighbours  || relative_rank(Us, s) >= RANK_5 || lever)

mcostalba commented:
Can you please leave relative_rank at the end?  Changes nothing speedwise but is more consistent.

Rocky640 commented:
OK, done

mcostalba commented:
@Rocky640  Thanks :-)

mcostalba commented:
**WAIT!!!!**   I found a major flow here!

neighbours can be anywhere compared to our square s, in particularly can be behind s, so that 

`     b = rank_bb(backmost_sq(Us, neighbours | stoppers));
`
 
Can return a rank behind 's' and this cannot be a backward pawn.

I have verified with this printing:
 
```
            // The pawn is backward when it cannot safely progress to that rank:
            // either there is a stopper in the way on this rank, or there is a
            // stopper on adjacent file which control the way to that rank.
            backward = (b | shift_bb<Up>(b & adjacent_files_bb(f))) & stoppers;

            if (backward && b < rank_bb(s))
                std::cerr << UCI::square(s) << "\n" << pos << std::endl;
```

mcostalba commented:
Please discard previous post, '& stoppers' condition ensures that only forward squares are considered.

Sorry for the noise :-(

Rocky640 commented:
Indeed, and also because we know that s is not connected, so if we have neighbours behind, it will be at least 2 ranks, so the shift_bb is still behind.

Rocky640 commented:
I was about to answer to your post something when I realised we do not need the connected condition either !!!

So amazingly, the following three lines, although not equivalent 
will always gives the same backward result at the end. 

Some will go in the else more often
a)
if (!neighbours || lever || relative_rank(Us, s) >= RANK_5)
b)
if (connected || !neighbours || lever || relative_rank(Us, s) >= RANK_5)
c) 
if (!neighbours || !(pawn_attack_span(Them, s + Up) & neighbours) || lever || relative_rank(Us, s) >= RANK_5)

It would just be a matter to find out which one is faster. 
But my measures are usually not so good.

mcostalba commented:
Nice catch!  I don't know which is faster, but I know which is simpler, and
is (a)

So of course I'd go with a  :-)



On Tue, Apr 5, 2016 at 2:16 PM, Rocky640 <notifications@github.com> wrote:

> I was about to answer to your post something when I realised we do not
> need the connected condition either !!!
>
> So amazingly, the following three lines, although not equivalent
> will always gives the same backward result at the end.
>
> Some will go in the else more often
> a)
> if (!neighbours || lever || relative_rank(Us, s) >= RANK_5)
> b)
> if (connected || !neighbours || lever || relative_rank(Us, s) >= RANK_5)
> c)
> if (!neighbours || !(pawn_attack_span(Them, s + Up) & neighbours) || lever
> || relative_rank(Us, s) >= RANK_5)
>
> It would just be a matter to find out which one is faster.
> But my measures are usually not so good.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/614#issuecomment-205773526>
>

Rocky640 commented:
LOL, I just push the simplified simplified, rearranged, simplified, fixed and simplified version.
I also added an assert to make clear that a pawn cannot be backward and have neighbours on same rank or below.

Still I would like someone to compare 
a) https://github.com/official-stockfish/Stockfish/pull/614/commits/6c1c1f3a6d5ba7ae6df73245c01a1a22e6e5b2b3

and the last one:
b) https://github.com/official-stockfish/Stockfish/pull/614/commits/0cdd1b2b80ea7b66d2fb802f2c7aa6880f2df302

because 
a) without connected, it goes in the "if" 73% of the time.
b) with "connected" it goes in the "if" only 17% of the time 

let's forget about 
c) with "pawn_attack_span(Them, s + Up) & neighbours", only 14% of the time, but too ugly.
    and b) is almost as good and also cheaper to compute.

FauziAkram commented:
After all these changes I think a new test should be made

anshulongithub commented:
Yes i also feel that a new test should be run before the final code is merged into master. Alain, will it be possible to test the final version again? Just to double confirm that its still better than the original version for which this was a simplification. Please advise.

Rocky640 commented:
The functional change (changing the backward definition) was tested on fishbench as reported at the very beginning of this pull request.
Since then, all changes made are "no functional changes", except the merge with latest master,
The bench has changed. only because of the merge with latest master.

My (personal) small concern is which one to chose, and for this a simple speed test will do.

mcostalba commented:
No need to retest IMO.


On Tue, Apr 5, 2016 at 8:54 PM, anshulongithub <notifications@github.com>
wrote:

> Yes i also feel that a new test should be run before the final code is
> merged into master. Alain, will it be possible to test the final version
> again? Just to double confirm that its still better than the original
> version for which this was a simplification. Please advise.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/official-stockfish/Stockfish/pull/614#issuecomment-205948138>
>

zamar commented:
approved.
